tools/bin/commands/dev

```diff
@@ -1,15 +1,102 @@
 #!/bin/bash
 
+COOKIE1="$HOME/.syndesis_api_cookie"
+
+
+# helper python command to urlencode
+dev::urlencode() {
+
+local py_script="
+import sys, urllib as ul
```

Comment:
can we please avoid using Python ? I really would like to keep it a plain bash script for portability reasons. 

We deliberately choose bash as platform as the least common denominator which everyone can work on. Actually would have loved to implement it in Perl or Golang as these are much more powerful languages.

Shell is painful, true. But please let keep the pain pure.

Reply:
Python is a pre-requisite on RHEL, since its package manager is written with that language, and is installed by default on MacOSX.
Not present on Windows, likewise bash.

An eventual additional package to implement urlencode from command line has way less chances to be installed on a platform than python.

golang is just not scripting, and I would never use it as a replacement for scripting

Reply:
I was afraid that you start arguing like that.

Ok, so I have to tell the long story. It's **not** about whether Python is installed or not. Perl is installed, too, everywhere. It's about respecting the team and allow as many people as possible to maintain this script.

When I started the script I seriously considered to write it in Perl because Perl is my mother language. I can do everything in Perl, super fast. "syndesis" would be much more powerful, we could do everything. 

But I would be the only one who could maintain it. So, after some discussion, e.g. with @iocanel, we bite in the the sour apple to implement everything in Shell. Not that I'm a big fan of shell scripting, not at all. But because *everyone* can understand it (or is supposed to at least). The challenge of course was to write sustainable shell script (where I learned quite some lessons, so many, that I'm even submitting CFP to conference for how to teach 'idiomatic shell' scripting).

For Python the same is true as for Perl. I don't know Python at all, at not many do. So you would be alone to maintain it.

So please, let's stick to plain shell. Don't open the door to Python or any other scripting language (why wouldn't we then have chosen Python from the beginning ?)

tl;dr - its not a technical reason, its a cultural and social reason to not introduce Python. I think we should be able to do it with plain UNIX commands.

(and what, btw, will happen if someone has Python 3 installed instead of Python 2 (e.g. with brew). Are you user urllib works here the same ? As they have removed urllib2 and are not really famous to be backwards comptabible ....)

Reply:
seems that urldecoding can be simply done in bash with `decoded=$(printf '%b' "${input//%/\\x}")` according to https://unix.stackexchange.com/questions/159253/decoding-url-encoding-percent-encoding

Reply:
* https://stackoverflow.com/a/10660730/207604 (not saying thats the best solution, just the first hit)

```
rawurlencode() {
  local string="${1}"
  local strlen=${#string}
  local encoded=""
  local pos c o

  for (( pos=0 ; pos<strlen ; pos++ )); do
     c=${string:$pos:1}
     case "$c" in
        [-_.~a-zA-Z0-9] ) o="${c}" ;;
        * )               printf -v o '%%%02x' "'$c"
     esac
     encoded+="${o}"
  done
  echo "${encoded}"    # You can either set a return variable (FASTER) 
  REPLY="${encoded}"   #+or echo the result (EASIER)... or both... :p
}
```

and now for the devils advocate argument, if not shell why not

```perl
value="$(perl -MURI::Escape -e 'print uri_escape($ARGV[0]);' "$1")"
```

Reply:
> seems that urldecoding can be simply done in bash with decoded=$(printf '%b' "${input//%/\\x}") according to unix.stackexchange.com/questions/159253/decoding-url-encoding-percent-encoding

I think its more about encoding then decoding. Though not sure why because you could always use `curl --data-urlencode`. @paoloantinori isn't that good enough when sending it back via curl ?